File: javascript/03-copy-paste-payroll/test/payroll_calculator.test.js
import assert from 'node:assert';
import test from 'node:test';
import { PayrollCalculator } from '../src/payroll_calculator.js';

test('calculates fulltime payslip', () => {
  const calculator = new PayrollCalculator();
  const result = calculator.generate_payslips([{ id: 1, name: 'Alice', type: 'fulltime', salary: 60000 }]);
  assert.strictEqual(result[0].gross, 5000);
  assert.strictEqual(result[0].deductions, 1250);
  assert.strictEqual(result[0].net, 3750);
});

test('calculates fulltime with bonus', () => {
  const calculator = new PayrollCalculator();
  const result = calculator.generate_payslips([{ id: 1, name: 'Alice', type: 'fulltime', salary: 60000, bonus: 12000 }]);
  assert.strictEqual(result[0].gross, 6000);
  assert.strictEqual(result[0].net, 4750);
});

test('calculates parttime payslip', () => {
  const calculator = new PayrollCalculator();
  const result = calculator.generate_payslips([{ id: 2, name: 'Bob', type: 'parttime', hours: 80, rate: 25 }]);
  assert.strictEqual(result[0].gross, 2000);
  assert.strictEqual(result[0].deductions, 300);
  assert.strictEqual(result[0].net, 1700);
});

test('calculates contract payslip', () => {
  const calculator = new PayrollCalculator();
  const result = calculator.generate_payslips([{ id: 3, name: 'Carol', type: 'contract', flatFee: 5000 }]);
  assert.strictEqual(result[0].gross, 5000);
  assert.strictEqual(result[0].deductions, 500);
  assert.strictEqual(result[0].net, 4500);
});

test('handles multiple employees', () => {
  const calculator = new PayrollCalculator();
  const result = calculator.generate_payslips([
    { id: 1, name: 'Alice', type: 'fulltime', salary: 60000 },
    { id: 2, name: 'Bob', type: 'parttime', hours: 80, rate: 25 }
  ]);
  assert.strictEqual(result.length, 2);
  assert.strictEqual(result[0].name, 'Alice');
  assert.strictEqual(result[1].name, 'Bob');
});

test('returns empty array for empty input', () => {
  const calculator = new PayrollCalculator();
  const result = calculator.generate_payslips([]);
  assert.deepStrictEqual(result, []);
});

File: csharp/03-copy-paste-payroll/test/payroll_calculator_test.cs
using Xunit;

public class PayrollCalculatorTest
{
    [Fact]
    public void CalculatesFulltimePayslip()
    {
        var calculator = new PayrollCalculator();
        var result = calculator.generate_payslips(new List<Employee>
        {
            new Employee { Id = 1, Name = "Alice", Type = "fulltime", Salary = 60000 }
        });
        Assert.Equal(5000m, result[0].Gross);
        Assert.Equal(1250m, result[0].Deductions);
        Assert.Equal(3750m, result[0].Net);
    }

    [Fact]
    public void CalculatesFulltimeWithBonus()
    {
        var calculator = new PayrollCalculator();
        var result = calculator.generate_payslips(new List<Employee>
        {
            new Employee { Id = 1, Name = "Alice", Type = "fulltime", Salary = 60000, Bonus = 12000 }
        });
        Assert.Equal(6000m, result[0].Gross);
        Assert.Equal(4750m, result[0].Net);
    }

    [Fact]
    public void CalculatesParttimePayslip()
    {
        var calculator = new PayrollCalculator();
        var result = calculator.generate_payslips(new List<Employee>
        {
            new Employee { Id = 2, Name = "Bob", Type = "parttime", Hours = 80, Rate = 25 }
        });
        Assert.Equal(2000m, result[0].Gross);
        Assert.Equal(300m, result[0].Deductions);
        Assert.Equal(1700m, result[0].Net);
    }

    [Fact]
    public void CalculatesContractPayslip()
    {
        var calculator = new PayrollCalculator();
        var result = calculator.generate_payslips(new List<Employee>
        {
            new Employee { Id = 3, Name = "Carol", Type = "contract", FlatFee = 5000 }
        });
        Assert.Equal(5000m, result[0].Gross);
        Assert.Equal(500m, result[0].Deductions);
        Assert.Equal(4500m, result[0].Net);
    }

    [Fact]
    public void HandlesMultipleEmployees()
    {
        var calculator = new PayrollCalculator();
        var result = calculator.generate_payslips(new List<Employee>
        {
            new Employee { Id = 1, Name = "Alice", Type = "fulltime", Salary = 60000 },
            new Employee { Id = 2, Name = "Bob", Type = "parttime", Hours = 80, Rate = 25 }
        });
        Assert.Equal(2, result.Count);
        Assert.Equal("Alice", result[0].Name);
        Assert.Equal("Bob", result[1].Name);
    }

    [Fact]
    public void ReturnsEmptyArrayForEmptyInput()
    {
        var calculator = new PayrollCalculator();
        var result = calculator.generate_payslips(new List<Employee>());
        Assert.Empty(result);
    }
}
